### PR TITLE
Replace boost::random

### PR DESCRIPTION
**Pull request by masterleinad**
Following #8310 and #8222, these are the changes necessary to replace `boost::random` by `std`.
While being there, I also unified the two `DoFRenumbering::compute_random` overloads.

### CODE REVIEW

**Comment by masterleinad** [comment]
Hmm... I can't reproduce the test failures with any of my setups...

**Comment by drwells** [comment]
Is there any way that we could match the seed that boost uses so that we get the same results with the `std` replacements?

**Comment by masterleinad** [comment]
> Is there any way that we could match the seed that boost uses so that we get the same results with the `std` replacements?

`std::mt19937(5489)` gives the same output as `boost::random::mt19937` and the `unit_int_distribution` versions are compatible as well. This doesn't seem to hold for `uniform_real_distribution`, though. 

Can you reproduce the test failures? Maybe, we really need alternative output files.

**Comment by drwells** [comment]
I will try to reproduce the test failures later today.

**Comment by drwells** [comment]
Yup, I can reproduce the failures with `dof_renumbering_08` and `dof_renumbering_09`. I will try to find a way to fix the output before we add alternative files.

**Comment by drwells** [comment]
It looks like the algorithms used in the standard and in boost to generate normally distributed numbers are different, so there is no straightforward way to get the same outputs.

I am generating the alternative outputs.

**Comment by masterleinad** [comment]
> I am generating the alternative outputs.

Thanks!

**Comment by drwells** [comment]
I added alternate output files that make the two failing tests pass for me (I think the serialization failure is spurious; we'll see).

What compiler did you use when you updated the output files?

**Comment by masterleinad** [comment]
> What compiler did you use when you updated the output files?

This is with `gcc-9.1.0`.

**Comment by drwells** [comment]
Weird.

I will force-push an updated second commit with the other output file (apparently we test serialization with a random mesh)

**Comment by drwells** [comment]
OK, that should fix it.

**Comment by masterleinad** [comment]
Finally!

**Comment by bangerth** [comment]
I don't have much of an opinion on whether we want to change. I think it boils down to how different compilers implement the `std::random` pieces. If we only have two tests that differ between compilers, I think that's acceptable.

File: include/deal.II/dofs/dof_renumbering.h [hunk 1]
```diff
@@ -1152,25 +1152,17 @@ namespace DoFRenumbering
...
-   * Compute the renumbering vector needed by the random() function. Same
-   * as the above function but for a single level of a multilevel
-   * discretization.
-   */
-  template <typename DoFHandlerType>
-  void
   compute_random(std::vector<types::global_dof_index> &new_dof_indices,
                  const DoFHandlerType &                dof_handler,
-                 const unsigned int                    level);
+                 const unsigned int level = numbers::invalid_unsigned_int);
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply include/deal.II/dofs/dof_renumbering.h]: My reasoning was just avoiding code duplication without any notable change (except for the documentation) for the user.

[the remark this answers is not in the document]
Comment (bangerth) [reply include/deal.II/dofs/dof_renumbering.h]: It's just a departure from otherwise common practice. I'm usually for consistency.

[the remark this answers is not in the document]
Comment (masterleinad) [reply include/deal.II/dofs/dof_renumbering.h]: I am not going to fight for this and will revert it then.

**Comment by drwells** [comment]
Lets give it another day and then merge if no one objects.

File: source/dofs/dof_renumbering.cc [hunk 2]
```diff
@@ -2131,26 +2130,19 @@ namespace DoFRenumbering
...
     const types::global_dof_index n_dofs = dof_handler.n_dofs();
+    (void)n_dofs;
     Assert(new_indices.size() == n_dofs,
            ExcDimensionMismatch(new_indices.size(), n_dofs));
 
-    for (unsigned int i = 0; i < n_dofs; ++i)
-      new_indices[i] = i;
+    std::iota(new_indices.begin(),
+              new_indices.end(),
+              types::global_dof_index(0));
```
Comment (bangerth) [inline source/dofs/dof_renumbering.cc]: Nice :-)

File: source/dofs/dof_renumbering.inst.in [hunk 3]
```diff
@@ -288,11 +288,6 @@ for (deal_II_dimension : DIMENSIONS; deal_II_space_dimension : SPACE_DIMENSIONS)
       template void
       random<DoFHandler<deal_II_dimension>>(DoFHandler<deal_II_dimension> &);
 
-      template void
-      compute_random<DoFHandler<deal_II_dimension>>(
-        std::vector<types::global_dof_index> &,
-        const DoFHandler<deal_II_dimension> &);
```
Comment (bangerth) [inline source/dofs/dof_renumbering.inst.in]: Don't you have to revert the changes to file as well?

File: tests/fe/nedelec_non_rect_2d.cc [hunk 4]
```diff
@@ -315,7 +315,7 @@ namespace polytest
     tria.refine_global(1);
 
     // DISTORT ALL:
-    GridTools::distort_random(0.2, tria, false);
+    // GridTools::distort_random(0.2, tria, false);
```
Comment (bangerth) [inline tests/fe/nedelec_non_rect_2d.cc]: Was this change on purpose?

File: source/dofs/dof_renumbering.inst.in (hunk 3 above)
Comment (masterleinad) [reply source/dofs/dof_renumbering.inst.in]: Of course! :pensive:

File: tests/fe/nedelec_non_rect_2d.cc (hunk 4 above)
Comment (masterleinad) [reply tests/fe/nedelec_non_rect_2d.cc]: Nope, fixed.

**Comment by bangerth** [comment]
Please squash the last two commits into the first (I think that's the most logical choice) and merge yourself!
